File: ADT/trie/212.cpp
// Google, Amazon, Yahoo, Microsoft, Meta, MakeMyTrip, Nvidia
// Tc = O(k·L + m·n·4^L)
// Sc = O(k·L + L)

class Solution {
public:
    vector<string> result;          // O(k) space for answers
    int r, c;
    
    // Constant directions → O(1)
    vector<pair<int, int>> directions{{-1, 0},{1, 0},{0, 1},{0, -1}};
    
    struct trieNode {
        bool endOfWord;
        trieNode* children[26];     // Constant size → O(1)
        string word;
    };

    // Creates one trie node → O(1)
    trieNode* getNode() {
        trieNode* temp = new trieNode();
        temp->endOfWord = false;
        
        // Always 26 iterations → O(1)
        for(int i = 0; i < 26; i++) {
            temp->children[i] = NULL;
        }
        
        temp->word = "";
        return temp;
    }

    // Insert one word of length L → O(L)
    void insert(trieNode* root, string str) {
        trieNode* pCrawl = root;
        
        // Loop runs for each character of word → O(L)
        for(char ch : str) {
            if(pCrawl->children[ch - 'a'] == NULL) {
                pCrawl->children[ch - 'a'] = getNode(); // O(1)
            }
            pCrawl = pCrawl->children[ch - 'a']; // O(1)
        }
        
        pCrawl->endOfWord = true;   // O(1)
        pCrawl->word = str;         // O(1)
    }

    // DFS depth ≤ max word length L
    void DFS(vector<vector<char>>& board, int i, int j, trieNode* root) {
        
        // Boundary + visited + trie pruning → O(1)
        if(i < 0 || i >= r || j < 0 || j >= c ||
           board[i][j] == '$' ||
           root->children[board[i][j] - 'a'] == NULL) {
            return;
        }

        // Move down the trie → O(1)
        root = root->children[board[i][j] - 'a'];

        // Found a word → O(1)
        if(root->endOfWord == true) {
            result.push_back(root->word); // amortized O(1)
            root->endOfWord = false;      // prevents duplicates
        }

        char temp = board[i][j];  // O(1)
        board[i][j] = '$';        // O(1)

        // Always 4 directions → O(4) = O(1)
        // BUT recursive calls branch → contributes to 4^L
        for(pair<int, int> p : directions) {
            DFS(board, i + p.first, j + p.second, root);
        }

        board[i][j] = temp;       // Backtracking → O(1)
    }

    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {

        r = board.size();
        c = board[0].size();

        trieNode* root = getNode();   // O(1)

        // Insert k words
        // Total = sum of lengths = O(k · L)
        for(string str : words) {
            insert(root, str);
        }

        // Loop over every cell → O(r · c)
        for(int i = 0; i < r; i++) {
            for(int j = 0; j < c; j++) {

                char ch = board[i][j];

                // O(1) check before DFS
                if(root->children[ch - 'a'] != NULL) {

                    // DFS worst-case → O(4^L)
                    DFS(board, i, j, root);
                }
            }
        }

        return result;
    }
};

File: ADT/tries/212.py
# Google, Microsoft, Yahoo, Apple
# Tc = O(m.k + r.c.4^k)
# Sc = O(m.k)

class TrieNode:
    def __init__(self):
        self.end_of_word = False
        self.children = {}
        self.word = ""

class Solution:
    def __init__(self):
        self.result = []
        self.r = 0
        self.c = 0
        self.directions = [(-1, 0), (1, 0), (0, 1), (0, -1)]

    def insert(self, root, word):
        pCrawl = root
        for ch in word:
            if ch not in pCrawl.children:
                pCrawl.children[ch] = TrieNode()
            pCrawl = pCrawl.children[ch]
        pCrawl.end_of_word = True
        pCrawl.word = word

    def DFS(self, board, i, j, root):
        if i < 0 or i >= self.r or j < 0 or j >= self.c or board[i][j] == '$' or board[i][j] not in root.children:
            return
        
        root = root.children[board[i][j]]
        if root.end_of_word:
            self.result.append(root.word)
            root.end_of_word = False  # avoid duplicates
        
        temp = board[i][j]
        board[i][j] = '$'  # mark as visited
        
        for direction in self.directions:
            new_i, new_j = i + direction[0], j + direction[1]
            self.DFS(board, new_i, new_j, root)
        
        board[i][j] = temp  # restore the original character

    def findWords(self, board, words):
        self.r = len(board)
        self.c = len(board[0]) if self.r > 0 else 0

        root = TrieNode()
        for word in words:
            self.insert(root, word)

        for i in range(self.r):
            for j in range(self.c):
                if board[i][j] in root.children:
                    self.DFS(board, i, j, root)

        return self.result
